src/diffusers/models/vae.py

```diff
@@ -280,6 +280,292 @@ def custom_forward(*inputs):
+    ) -> None:
+        super().__init__()
+        self.in_channels = in_channels
+        self.out_channels = out_channels
+        self.activate_before = activate_before
+        self.activate_after = activate_after
+        self.upsample_type = upsample_type
+        if self.upsample_type == "deconv":
+            self.deconv = nn.ConvTranspose2d(in_channels, out_channels, kernel_size=4, stride=2, padding=1)
+        else:
+            if self.upsample_type in ["bilinear", "nearest"]:
+                self.conv = nn.Conv2d(in_channels, out_channels, kernel_size=3, stride=1, padding=1)
+            else:
+                raise NotImplementedError
+
+    def forward(self, x: torch.FloatTensor) -> torch.FloatTensor:
+        if self.activate_before == "relu":
+            x = torch.relu(x)
+        elif self.activate_before == "none":
+            pass
+        else:
+            raise NotImplementedError
+
+        if self.upsample_type == "deconv":
+            x = self.deconv(x)
+        else:
+            x = nn.functional.interpolate(x, scale_factor=2.0, mode=self.upsample_type)
+            x = self.conv(x)
+
+        if self.activate_after == "relu":
+            x = torch.relu(x)
+        elif self.activate_after == "none":
+            pass
+        else:
+            raise NotImplementedError
+
+        return x
+
+
+class ConditionEncoder(nn.Module):
+    """
+    used in AsymmetricAutoencoderKL
+    """
+
+    def __init__(
+        self,
+        in_ch: int,
+        out_ch: int = 192,
+        res_ch: int = 768,
+        stride: int = 16,
+        upsample_type: str = "deconv",
+        downsample_layer: str = "conv",
+    ) -> None:
+        super().__init__()
+
+        up_layers = []
+        while stride > 1:
+            stride = stride // 2
+            in_ch_ = out_ch * 2
+            if out_ch > res_ch:
+                out_ch = res_ch
+            if stride == 1:
+                in_ch_ = res_ch
+            up_layers.append(
+                UpSample(in_ch_, out_ch, activate_before="relu", activate_after="none", upsample_type=upsample_type)
+            )
+            out_ch *= 2
+        self.up_layers = nn.Sequential(*reversed(up_layers))
+
+        out_channels = []
+        for layer in up_layers:
+            out_channels.append(layer.out_channels)
+        out_channels.append(up_layers[-1].in_channels)
+
+        layers = []
+        in_ch_ = in_ch
+        for l in range(len(out_channels)):
+            out_ch_ = out_channels[l]
+            if l == 0 or l == 1:
+                layers.append(nn.Conv2d(in_ch_, out_ch_, kernel_size=3, stride=1, padding=1))
+            else:
+                if downsample_layer == "conv":
+                    layers.append(nn.Conv2d(in_ch_, out_ch_, kernel_size=4, stride=2, padding=1))
+                else:
+                    raise NotImplementedError
+            in_ch_ = out_ch_
+
+        self.layers = nn.Sequential(*layers)
+        self.downsample_layer = downsample_layer
```

Comment:
Can we try to clean this a bit? If I read correctly, we don't need the `self.up_layers` at all